BUG: Use unrotated companion matrix in polynomial.polyroots.

ForestCamper opened the pull request:
Fixes #27881.

Both functions use numpy.linalg.eigvals, but while roots gives in argument the polynomial's companion matrix unchanged, polyroots rotates it.

https://github.com/numpy/numpy/blob/5570e92f8b604ab725a8a605221febf49e168629/numpy/polynomial/polynomial.py#L1538-L1539

Though in theory this rotation shouldn't change anything, in terms of numerical calcuations, eigvals gives significantly different results. This PR removes the rotation as an easy fix to the inconsistency.

This strange behavior by eigvals is however a bug. I did some research on it, which you can find on this [issue](https://github.com/numpy/numpy/issues/27881), but I haven't any solution.

<!--         ----------------------------------------------------------------
                MAKE SURE YOUR PR GETS THE ATTENTION IT DESERVES!
                ----------------------------------------------------------------

*  FORMAT IT RIGHT:
      https://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message

*  IF IT'S A NEW FEATURE OR API CHANGE, TEST THE WATERS:
      https://www.numpy.org/devdocs/dev/development_workflow.html#get-the-mailing-list-s-opinion

*  HIT ALL THE GUIDELINES:
      https://numpy.org/devdocs/dev/index.html#guidelines

*  WHAT TO DO IF WE HAVEN'T GOTTEN BACK TO YOU:
      https://www.numpy.org/devdocs/dev/development_workflow.html#getting-your-pr-reviewed
-->

ForestCamper committed:
BUG: Fix #27881 inconsistent behavior between numpy.roots and numpy.polynomial.polynomial.polyroots

Both functions use numpy.linalg.eigvals, but while roots gives in argument the polynomial's companion matrix unchanged, polyroots rotates it. Though in theory this rotation shouldn't change anything, in terms of numerical calcuations, eigvals gives significantly different results. This commit removes the rotation as an easy fix to the inconsistency.

This strange behavior by eigvals is however a bug. I did some research on it, which you can find on the issue.

ngoldbaum commented:
This needs at least some tests that demonstrate what this change is doing. Given that you made this change and CI passed, that indicates our test coverage is poor here. It would help to improve confidence that this is a good change if you also improved test coverage alongside the change.

ForestCamper committed:
BUG: Fix #27881 Adding tests for numpy.roots and numpy.polynomial.polynomial.polyroots

The following tests provide a minimum precision expected by the functions. In the case of my change, they allow my change but not the previous version. You'll find that the difference of precision between the two versions vary according to the root values but in all cases, this change either increases precision or maintains the previous result.

ForestCamper committed:
BUG: Fix #27881 Adding tests for numpy.roots and numpy.polynomial.polynomial.polyroots

Fixing tests

ForestCamper committed:
BUG: Fix #27881 Adding tests for numpy.roots and numpy.polynomial.polynomial.polyroots

Fixing tests. It seems that the way numpy.roots creates it's matrice companion augments the numerical error compared to the polycompanion function, which is why I had to reduce the expected precision.

charris commented:
I think this is a numerical corner case and not to worry about, but using the rotated companion matrix was also playing to a corner case. So let's make the change. I expect the algorithm internals have also changed over the years.

As a note, I wrote the original.

charris commented:
@jorenham Thoughts on the `*.pyi` changes?

jorenham commented on numpy/matlib.pyi:
what's the reason behind this change?

charris commented:
> That's a mistake, I don't know how it got there

You should fix that then.

ForestCamper committed:
Reverting change from commit 6703b91a22f7f0d1e98f565efcb659d8369b4952 in matlib.pyi

charris commented:
Thanks @Chevali2004 .
